File: backend_medjet/models/AttendanceDeviceModel.php
<?php

/**
 * Biometric attendance terminals (ZKTeco ADMS "push" devices).
 *
 * A device row exists in one of two states:
 *   - unclaimed: the terminal has contacted us but no company has entered its
 *     serial number yet. Carries no tenant data.
 *   - active/disabled: claimed by exactly one tenant and bound to one branch.
 *
 * Claiming is first-come-first-served on the serial number, which is printed
 * on the device — whoever physically holds it owns it.
 */
final class AttendanceDeviceModel {
    /** Serial numbers are compared upper-cased and trimmed everywhere. */
    public static function normaliseSerial($raw): string {
        return strtoupper(trim((string) $raw));
    }

    public static function findBySerial(string $serial): ?array {
        return Database::fetchOne(
            "SELECT * FROM attendance_devices WHERE serial_number = ? LIMIT 1",
            [self::normaliseSerial($serial)]
        );
    }

    public static function findById(int $id, int $tenantId): ?array {
        return Database::fetchOne(
            "SELECT * FROM attendance_devices WHERE id = ? AND tenant_id = ? LIMIT 1",
            [$id, $tenantId]
        );
    }

    /**
     * The stand-in device for file imports at a branch, created on first use.
     *
     * A punch has to belong to a device row — that is what carries the branch,
     * the clock offset and the repeat-tap window. A customer importing a CSV
     * usually has no registered terminal at all (that is why they are importing
     * a file), so one is synthesised per branch.
     *
     * The serial is namespaced and unclaimable: `normaliseSerial` upper-cases,
     * and register.php rejects anything that is not `[A-Z0-9-]{4,64}`, so a real
     * terminal can never collide with `FILE-T3-B7` — and nobody can type it in
     * to hijack an import stream, because these ids are per tenant already.
     */
    public static function ensureFileImportDevice(int $tenantId, int $branchId, ?int $adminId): array {
        $serial = self::normaliseSerial("FILE-T{$tenantId}-B{$branchId}");
        $existing = self::findBySerial($serial);
        if ($existing !== null) {
            return $existing;
        }

        Database::execute(
            "INSERT INTO attendance_devices
                (tenant_id, branch_id, serial_number, name, vendor, status, claimed_by, claimed_at, first_seen_at)
             VALUES (?, ?, ?, ?, 'other', 'active', ?, NOW(), NOW())
             ON DUPLICATE KEY UPDATE id = id",
            [$tenantId, $branchId, $serial, 'استيراد ملف', $adminId]
        );

        return self::findBySerial($serial);
    }

    /**
     * Records that a serial number just talked to us, creating the unclaimed
     * row on first contact. Returns the device row.
     *
     * Called on the hot device path, so it stays a single upsert.
     */
    public static function recordContact(string $serial, ?string $ip): array {
        $serial = self::normaliseSerial($serial);

        Database::execute(
            "INSERT INTO attendance_devices (serial_number, status, first_seen_at, last_seen_at, last_ip)
             VALUES (?, 'unclaimed', NOW(), NOW(), ?)
             ON DUPLICATE KEY UPDATE last_seen_at = NOW(), last_ip = VALUES(last_ip)",
            [$serial, $ip]
        );

        $device = self::findBySerial($serial);
        if ($device === null) {
            // Only reachable if the row was deleted between the two statements.
            throw new RuntimeException('Device row vanished for serial ' . $serial);
        }
        return $device;
    }

    /** Stores whatever the device volunteered about itself during the handshake. */
    public static function updateInfo(int $deviceId, array $info): void {
        $sets = [];
        $params = [];
        foreach (['model' => 'model', 'firmware' => 'firmware', 'user_count' => 'user_count'] as $key => $column) {
            if (array_key_exists($key, $info) && $info[$key] !== null && $info[$key] !== '') {
                $sets[] = "`$column` = ?";
                $params[] = $info[$key];
            }
        }
        if (!$sets) {
            return;
        }
        $params[] = $deviceId;
        Database::execute(
            "UPDATE attendance_devices SET " . implode(', ', $sets) . " WHERE id = ?",
            $params
        );
    }

    public static function touchPunch(int $deviceId, string $punchedAt): void {
        Database::execute(
            "UPDATE attendance_devices
             SET last_punch_at = GREATEST(COALESCE(last_punch_at, ?), ?)
             WHERE id = ?",
            [$punchedAt, $punchedAt, $deviceId]
        );
    }

    /**
     * Binds an unclaimed device to a tenant + branch.
     *
     * A device already owned by ANOTHER tenant is refused: silently moving it
     * would hand that company's punch stream to a stranger.
     */
    public static function claim(
        string $serial,
        int $tenantId,
        int $branchId,
        ?string $name,
        ?int $adminId
    ): array {
        $serial = self::normaliseSerial($serial);
        $existing = self::findBySerial($serial);

        if ($existing !== null
            && $existing['tenant_id'] !== null
            && (int) $existing['tenant_id'] !== $tenantId) {
            Response::fail(
                'This device is already registered to another company',
                409,
                'DEVICE_ALREADY_CLAIMED'
            );
        }

        if ($existing === null) {
            // Registered before the terminal ever dialled in. Legitimate — HR
            // usually types the serial off the box while the electrician is
            // still mounting it — so the row is created ready and waiting.
            Database::execute(
                "INSERT INTO attendance_devices
                    (tenant_id, branch_id, serial_number, name, status, claimed_by, claimed_at)
                 VALUES (?, ?, ?, ?, 'active', ?, NOW())",
                [$tenantId, $branchId, $serial, $name, $adminId]
            );
        } else {
            Database::execute(
                "UPDATE attendance_devices
                 SET tenant_id = ?, branch_id = ?, name = COALESCE(?, name),
                     status = 'active', claimed_by = ?, claimed_at = NOW()
                 WHERE id = ?",
                [$tenantId, $branchId, $name, $adminId, $existing['id']]
            );
        }

        return self::findBySerial($serial);
    }

    /**
     * Devices with everything the list screen needs to answer "is it working?".
     *
     * The freshness of `last_seen_at` is computed in SQL on purpose: those
     * timestamps are written with MySQL's NOW() (company local time) while PHP
     * runs in UTC, so comparing them in PHP would be three hours wrong.
     */
    public static function listForTenant(int $tenantId): array {
        return Database::fetchAll(
            "SELECT d.*, b.name AS branch_name,
                    TIMESTAMPDIFF(SECOND, d.last_seen_at, NOW()) AS seconds_since_seen,
                    (SELECT COUNT(*) FROM device_users du
                      WHERE du.device_id = d.id AND du.employee_id IS NOT NULL) AS linked_users,
                    (SELECT COUNT(*) FROM device_users du
                      WHERE du.device_id = d.id AND du.employee_id IS NULL) AS pending_users,
                    (SELECT COUNT(*) FROM device_punches p
                      WHERE p.device_id = d.id AND DATE(p.punched_at) = CURDATE()) AS punches_today
             FROM attendance_devices d
             LEFT JOIN branches b ON b.id = d.branch_id AND b.tenant_id = d.tenant_id
             WHERE d.tenant_id = ?
             ORDER BY d.id DESC",
            [$tenantId]
        );
    }

    public static function update(int $deviceId, int $tenantId, array $fields): void {
        $allowed = [
            'name', 'branch_id', 'status', 'direction_mode',
            'min_interval_seconds', 'clock_offset_minutes', 'keep_unmatched', 'debug_logging',
        ];
        $sets = [];
        $params = [];
        foreach ($allowed as $column) {
            if (array_key_exists($column, $fields)) {
                $sets[] = "`$column` = ?";
                $params[] = $fields[$column];
            }
        }
        if (!$sets) {
            return;
        }
        $params[] = $deviceId;
        $params[] = $tenantId;
        Database::execute(
            "UPDATE attendance_devices SET " . implode(', ', $sets) . " WHERE id = ? AND tenant_id = ?",
            $params
        );
    }

    /**
     * Releases a device back to unclaimed so it can be moved to another company
     * (resale, wrong company entered). The punch history stays: it belongs to
     * the company that recorded it, not to the hardware.
     */
    public static function release(int $deviceId, int $tenantId): void {
        Database::execute(
            "UPDATE attendance_devices
             SET tenant_id = NULL, branch_id = NULL, status = 'unclaimed',
                 claimed_by = NULL, claimed_at = NULL, name = NULL
             WHERE id = ? AND tenant_id = ?",
            [$deviceId, $tenantId]
        );
        Database::execute("DELETE FROM device_users WHERE device_id = ? AND tenant_id = ?", [$deviceId, $tenantId]);
        Database::execute("DELETE FROM device_commands WHERE device_id = ? AND tenant_id = ?", [$deviceId, $tenantId]);
    }

    /** True once the terminal has contacted us within the last few minutes. */
    public static function isOnline(?string $lastSeenAt, int $graceSeconds = 300): bool {
        if (!$lastSeenAt) {
            return false;
        }
        $row = Database::fetchOne(
            "SELECT TIMESTAMPDIFF(SECOND, ?, NOW()) AS age",
            [$lastSeenAt]
        );
        return $row !== null && (int) $row['age'] <= $graceSeconds;
    }
}

File: front_end/medjat_central/lib/data/model/attendance_device_model.dart
// A fingerprint / face terminal installed at a branch, the User IDs it knows
// about, and the raw punches it has pushed up.

class AttendanceDeviceModel {
  final int id;
  final String serialNumber;
  final String? name;
  final int? branchId;
  final String? branchName;
  final String? model;
  final String? firmware;
  final String status; // active | disabled
  final bool isOnline;
  final int? secondsSinceSeen;
  final String? lastSeenAt;
  final String? lastPunchAt;
  final String directionMode; // auto | device_status
  final int minIntervalSeconds;
  final int clockOffsetMinutes;
  final bool debugLogging;
  final int linkedUsers;
  final int pendingUsers;
  final int punchesToday;

  const AttendanceDeviceModel({
    required this.id,
    required this.serialNumber,
    this.name,
    this.branchId,
    this.branchName,
    this.model,
    this.firmware,
    this.status = 'active',
    this.isOnline = false,
    this.secondsSinceSeen,
    this.lastSeenAt,
    this.lastPunchAt,
    this.directionMode = 'auto',
    this.minIntervalSeconds = 60,
    this.clockOffsetMinutes = 0,
    this.debugLogging = false,
    this.linkedUsers = 0,
    this.pendingUsers = 0,
    this.punchesToday = 0,
  });

  String get displayName =>
      (name != null && name!.trim().isNotEmpty) ? name! : serialNumber;

  bool get isDisabled => status == 'disabled';

  /// True when the terminal has never contacted the server — the serial was
  /// typed in but the device's own network settings still need doing.
  bool get neverConnected => lastSeenAt == null;

  factory AttendanceDeviceModel.fromJson(Map<String, dynamic> json) {
    return AttendanceDeviceModel(
      id: (json['id'] as num?)?.toInt() ?? 0,
      serialNumber: json['serial_number']?.toString() ?? '',
      name: json['name']?.toString(),
      branchId: (json['branch_id'] as num?)?.toInt(),
      branchName: json['branch_name']?.toString(),
      model: json['model']?.toString(),
      firmware: json['firmware']?.toString(),
      status: json['status']?.toString() ?? 'active',
      isOnline: json['is_online'] == true,
      secondsSinceSeen: (json['seconds_since_seen'] as num?)?.toInt(),
      lastSeenAt: json['last_seen_at']?.toString(),
      lastPunchAt: json['last_punch_at']?.toString(),
      directionMode: json['direction_mode']?.toString() ?? 'auto',
      minIntervalSeconds: (json['min_interval_seconds'] as num?)?.toInt() ?? 60,
      clockOffsetMinutes: (json['clock_offset_minutes'] as num?)?.toInt() ?? 0,
      debugLogging: json['debug_logging'] == true,
      linkedUsers: (json['linked_users'] as num?)?.toInt() ?? 0,
      pendingUsers: (json['pending_users'] as num?)?.toInt() ?? 0,
      punchesToday: (json['punches_today'] as num?)?.toInt() ?? 0,
    );
  }
}

/// One User ID stored on the terminal, and the employee it points at.
class DeviceUserModel {
  final int id;
  final String deviceUserId;
  final String? deviceName;
  final int? employeeId;
  final String? employeeName;
  final String? employeeJobTitle;
  final String? cardNumber;
  final bool isDeviceAdmin;
  final String? lastPunchAt;
  final int unmatchedPunches;

  const DeviceUserModel({
    required this.id,
    required this.deviceUserId,
    this.deviceName,
    this.employeeId,
    this.employeeName,
    this.employeeJobTitle,
    this.cardNumber,
    this.isDeviceAdmin = false,
    this.lastPunchAt,
    this.unmatchedPunches = 0,
  });

  bool get isLinked => employeeId != null;

  factory DeviceUserModel.fromJson(Map<String, dynamic> json) {
    return DeviceUserModel(
      id: (json['id'] as num?)?.toInt() ?? 0,
      deviceUserId: json['device_user_id']?.toString() ?? '',
      deviceName: json['device_name']?.toString(),
      employeeId: (json['employee_id'] as num?)?.toInt(),
      employeeName: json['employee_name']?.toString(),
      employeeJobTitle: json['employee_job_title']?.toString(),
      cardNumber: json['card_number']?.toString(),
      isDeviceAdmin: json['is_device_admin'] == true,
      lastPunchAt: json['last_punch_at']?.toString(),
      unmatchedPunches: (json['unmatched_punches'] as num?)?.toInt() ?? 0,
    );
  }
}

/// A raw punch as the terminal sent it, plus what we did with it.
class DevicePunchModel {
  final int id;
  final int deviceId;
  final String? deviceName;
  final String deviceUserId;
  final String? deviceUserName;
  final int? employeeId;
  final String? employeeName;
  final String punchedAt;
  final String? direction; // in | out
  final String state; // applied | duplicate | unmatched | ignored | failed
  final String? note;
  final String recognition;

  const DevicePunchModel({
    required this.id,
    required this.deviceId,
    required this.deviceUserId,
    required this.punchedAt,
    required this.state,
    this.deviceName,
    this.deviceUserName,
    this.employeeId,
    this.employeeName,
    this.direction,
    this.note,
    this.recognition = 'device_fingerprint',
  });

  factory DevicePunchModel.fromJson(Map<String, dynamic> json) {
    return DevicePunchModel(
      id: (json['id'] as num?)?.toInt() ?? 0,
      deviceId: (json['device_id'] as num?)?.toInt() ?? 0,
      deviceName: json['device_name']?.toString(),
      deviceUserId: json['device_user_id']?.toString() ?? '',
      deviceUserName: json['device_user_name']?.toString(),
      employeeId: (json['employee_id'] as num?)?.toInt(),
      employeeName: json['employee_name']?.toString(),
      punchedAt: json['punched_at']?.toString() ?? '',
      direction: json['direction']?.toString(),
      state: json['state']?.toString() ?? 'unmatched',
      note: json['note']?.toString(),
      recognition: json['recognition']?.toString() ?? 'device_fingerprint',
    );
  }
}
